python/cugraph/traversal/bfs_wrapper.pyx

```diff
@@ -17,72 +17,88 @@
+from libc.stdlib cimport calloc, malloc, free
+from libc.float cimport FLT_MAX_EXP
 
 import cudf
 import cudf._lib as libcudf
+import rmm
 import numpy as np
 
-
 def bfs(input_graph, start, directed=True):
     """
     Call bfs
     """
-    cdef uintptr_t graph = graph_wrapper.allocate_cpp_graph()
-    cdef Graph * g = <Graph*> graph
+    # Step 1: Declare the different varibales
+    cdef GraphCSR[int, int, float]  graph_float     # For weighted float graph (SSSP) and Unweighted (BFS)
+    cdef GraphCSR[int, int, double] graph_double    # For weighted double graph (SSSP)
+
+    # Pointers required for CSR Graph
+    cdef uintptr_t c_offsets_ptr        = <uintptr_t> NULL # Pointer to the CSR offsets
+    cdef uintptr_t c_indices_ptr        = <uintptr_t> NULL # Pointer to the CSR indices
+
+    # Pointers for SSSP / BFS
+    cdef uintptr_t c_identifier_ptr     = <uintptr_t> NULL # Pointer to the DataFrame 'vertex' Series
+    cdef uintptr_t c_distance_ptr       = <uintptr_t> NULL # Pointer to the DataFrame 'distance' Series
+    cdef uintptr_t c_predecessor_ptr    = <uintptr_t> NULL # Pointer to the DataFrame 'predecessor' Series
 
-    if input_graph.adjlist:
-        [offsets, indices] = graph_wrapper.datatype_cast([input_graph.adjlist.offsets, input_graph.adjlist.indices], [np.int32])
-        [weights] = graph_wrapper.datatype_cast([input_graph.adjlist.weights], [np.float32, np.float64])
-        graph_wrapper.add_adj_list(graph, offsets, indices, weights)
-    else:
-        [src, dst] = graph_wrapper.datatype_cast([input_graph.edgelist.edgelist_df['src'], input_graph.edgelist.edgelist_df['dst']], [np.int32])
-        if input_graph.edgelist.weights:
-            [weights] = graph_wrapper.datatype_cast([input_graph.edgelist.edgelist_df['weights']], [np.float32, np.float64])
-            graph_wrapper.add_edge_list(graph, src, dst, weights)
-        else:
-            graph_wrapper.add_edge_list(graph, src, dst)
-        add_adj_list(g)
-        offsets, indices, values = graph_wrapper.get_adj_list(graph)
-        input_graph.adjlist = input_graph.AdjList(offsets, indices, values)
+    # Step 2: Verifiy input_graph has the expected format
+    if input_graph.adjlist is None:
+        input_graph.view_adj_list()
 
-    # we should add get_number_of_vertices() to Graph (and this should be
-    # used instead of g.adjList.offsets.size - 1)
-    num_verts = g.adjList.offsets.size - 1
+    # Step 3: Extract CSR offsets, indices, weights are not expected
+    #         - offsets: int (signed, 32-bit)
+    #         - indices: int (signed, 32-bit)
+    [offsets, indices] = graph_wrapper.datatype_cast([input_graph.adjlist.offsets, input_graph.adjlist.indices], [np.int32])
+    c_offsets_ptr = offsets.__cuda_array_interface__['data'][0]
+    c_indices_ptr = indices.__cuda_array_interface__['data'][0]
 
+    # Step 4: Setup number of vertices and edges
+    num_verts = input_graph.number_of_vertices()
+    num_edges = len(indices)
```

Comment:
input_graph.number_of_vertices()
```suggestion
    num_edges = input_graph.number_of_edges()
```

Reply:
This change could lead to errors for some cases. In case the user created a Graph() type with edgelist, input_graph.number_of_edges() gives the number of undirected edges now.

Reply:
Got it.